src/base/path.h

```diff
@@ -0,0 +1,108 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2021  Vladimir Golovnev <glassez@yandex.ru>
+ * Copyright (C) 2012  Christophe Dumez <chris@qbittorrent.org>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <QMetaType>
+#include <QString>
+
+class Path final
+{
+public:
+    enum class Format
+    {
+        Uniform,
+        Native
+    };
+
+    Path() = default;
+    Path(const Path &other) = default;
+    Path(Path &&other) = default;
+
+    explicit Path(const QString &pathStr);
+    explicit Path(const char pathStr[]);
+    explicit Path(const std::string &pathStr);
+
+    bool isValid() const;
+    bool isEmpty() const;
+    bool isAbsolute() const;
+    bool isRelative() const;
+
+    bool exists() const;
+
+    Path rootItem() const;
+    Path parentPath() const;
+    Path filename() const;
+
+    QString extension() const;
+    bool hasExtension(const QString &ext, Qt::CaseSensitivity caseSensitivity = Qt::CaseSensitive) const;
+    void removeExtension();
+
+    bool isEqualTo(const Path &other, Qt::CaseSensitivity caseSensitivity = Qt::CaseSensitive) const;
+    bool isBaseOf(const Path &other, Qt::CaseSensitivity caseSensitivity = Qt::CaseSensitive) const;
+    Path toRelativePath(const Path &basePath) const;
+
+    QString toString() const;
+    QString toNative() const;
+
+    Path &operator=(const Path &other) = default;
+    Path &operator=(Path &&other) = default;
+    bool operator==(const Path &other) const;
+    bool operator!=(const Path &other) const;
+
+    Path operator/(const Path &other) const;
```

Comment:
Those operators without assignment (`/`, `+`) should be out of class if possible. Or at least make them `friend`.